app/services/district_report_service.rb

```diff
@@ -0,0 +1,91 @@
+class DistrictReportService
+  def initialize(facilities:, selected_date:)
+    @facilities = Array(facilities)
+    @selected_date = selected_date
+    @data = {
+      controlled_patients: {},
+      registrations: {},
+      cumulative_registrations: 0,
+      quarterly_registrations: []
+    }.with_indifferent_access
+  end
+
+  attr_reader :selected_date, :facilities, :data
+
+  def call
+    compile_control_and_registration_data
+
+    compile_cohort_trend_data
+
+    data
+  end
+
+  def compile_control_and_registration_data
+    beginning_of_period = selected_date.advance(months: -11)
+    previous_registrations = 0
+    registrations.each do |(period, count)|
+      year, month = period[0].to_i, period[1].to_i
+      break if year > beginning_of_period.year
+      break if year == beginning_of_period.year && month >= beginning_of_period.month
+      previous_registrations += count
+    end
+
+    -11.upto(0).each do |n|
+      time = selected_date.advance(months: n)
+      formatted_period = time.to_s(:month_year)
+      key = [time.year.to_s, time.month.to_s]
+
+      @data[:controlled_patients][formatted_period] = controlled_patients(time).count
+      @data[:cumulative_registrations] ||= previous_registrations
+      @data[:cumulative_registrations] = @data[:cumulative_registrations] + registrations.fetch(key, 0).to_i.round
+      @data[:registrations][formatted_period] = @data[:cumulative_registrations]
+    end
+  end
+
+  def format_quarter(quarter)
+    "Q#{quarter.number}-#{quarter.year}"
+  end
+
+  # We want to return cohort data for the current quarter for the selected date, and then
+  # the previous three quarters. Each quarter cohort is made up of patients registered
+  # in the previous quarter who has had a follow up visit in the current quarter.
+  def compile_cohort_trend_data
```

Comment:
Not sure this code is necessary; see `CohortAnalyticsQuery` and `Analytics::DistrictsController` for how we compute these in the old dashboards